manager/orchestrator/global_test.go

```diff
@@ -214,13 +215,91 @@ func TestRemoveTask(t *testing.T) {
 	assert.Equal(t, observedTask2.NodeID, "id1")
 }
 
+func TestModeChange(t *testing.T) {
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
+
+	SetupCluster(t, s)
+
+	// start replicated orchestrator
+	replicated := New(s)
+	go func() {
+		assert.NoError(t, replicated.Run(context.Background()))
+	}()
+
+	watch, cancel := state.Watch(s.WatchQueue())
+	defer cancel()
+
+	skipEvents(t, watch)
+
+	var service *api.Service
+	s.View(func(tx store.ReadTx) {
+		service = store.GetService(tx, "id1")
+	})
+	assert.NotNil(t, service)
+
+	// change mode to replicated
+	service.Spec.Mode = &api.ServiceSpec_Replicated{
+		Replicated: &api.ReplicatedService{
+			Replicas: 1,
+		},
+	}
+	updateService(t, s, service)
+
+	observedTask := watchShutdownTask(t, watch)
+	assert.Equal(t, observedTask.ServiceAnnotations.Name, "name1")
+	assert.Equal(t, observedTask.NodeID, "id1")
+	assert.True(t, observedTask.Slot == 0)
+	assert.Equal(t, observedTask.DesiredState, api.TaskStateShutdown)
+
+	observedTask = watchTaskCreate(t, watch)
+	assert.Equal(t, observedTask.Status.State, api.TaskStateNew)
+	assert.Equal(t, observedTask.ServiceAnnotations.Name, "name1")
+	assert.True(t, observedTask.Slot == 1)
+
+	// change from replicated to global
+	s.View(func(tx store.ReadTx) {
+		service = store.GetService(tx, "id1")
+	})
+	assert.NotNil(t, service)
+
+	service.Spec.Mode = &api.ServiceSpec_Global{
+		Global: &api.GlobalService{},
+	}
+	updateService(t, s, service)
+
+	// give orchestrator time to update tasks
+	time.Sleep(50 * time.Millisecond)
```

Comment:
We need to avoid using sleeps like this in tests. It's easily for a loaded cloud VM to take longer than 50 ms to do something that normally would be instant.

The right way to do this is to start a watch and wait for the expected events.